src/machine/machine_gameboyadvance.go

```diff
@@ -18,27 +121,170 @@ func (p Pin) Set(value bool) {
 	// do nothing
 }
 
-var Display = FramebufDisplay{(*[160][240]volatile.Register16)(unsafe.Pointer(uintptr(0x06000000)))}
+// Displays is a convenience container for the various mode-based displays.
+type Displays struct {
+	Reg *volatile.Register16
+}
+
+// DISPCNT Register Constants
+const (
+	DISPCNT_MODE1 = 1
+	DISPCNT_MODE2 = 2
+	DISPCNT_MODE3 = 3
+	DISPCNT_MODE4 = 4
+	DISPCNT_MODE5 = 5
+
+	DISPCNT_FRAME_SELECT         = 1 << 4
+	DISPCNT_HBLANK_INTERVAL_FREE = 1 << 5
+	DISPCNT_OBJ_VRAM_MAP_1D      = 1 << 6
+	DISPCNT_FORCED_BLANK         = 1 << 7
+	DISPCNT_DISPLAY_BG0          = 1 << 8
+	DISPCNT_DISPLAY_BG1          = 1 << 9
+	DISPCNT_DISPLAY_BG2          = 1 << 10
+	DISPCNT_DISPLAY_BG3          = 1 << 11
+	DISPCNT_DISPLAY_OBJ          = 1 << 12
+
+	DISPCNT_MODE_MASK = 0x7
+	DISPCNT_BG_MASK   = 0xf << 8
+)
+
+type DisplayOption struct{ Mask, Set uint16 }
+
+var (
+	// Display modes (mutually exclusive)
+	Mode3 = DisplayOption{
+		// BG2 must be enabled in bitmapped modes, so set both at once.
+		Mask: DISPCNT_MODE_MASK | DISPCNT_BG_MASK,
+		Set:  DISPCNT_MODE3 | DISPCNT_DISPLAY_BG2,
+	}
+
+	// Enabled backgrounds (can be combined in one call to Configure)
+	Background0 = DisplayOption{Mask: DISPCNT_BG_MASK, Set: DISPCNT_DISPLAY_BG0}
+	Background1 = DisplayOption{Mask: DISPCNT_BG_MASK, Set: DISPCNT_DISPLAY_BG1}
+	Background2 = DisplayOption{Mask: DISPCNT_BG_MASK, Set: DISPCNT_DISPLAY_BG2}
+	Background3 = DisplayOption{Mask: DISPCNT_BG_MASK, Set: DISPCNT_DISPLAY_BG3}
+
+	// Display blanking (blanked at reset, mutually exclusive)
+	Blank   = DisplayOption{Set: DISPCNT_FORCED_BLANK}
+	Unblank = DisplayOption{Mask: DISPCNT_FORCED_BLANK}
+
+	// DisplayDefaults is useful for starter programs, but not much else.
+	DisplayDefaults = []DisplayOption{
+		Mode3,   // Enable 8bpp 160x240, single-frame bitmap mode
+		Unblank, // Start showing the display
+	}
+)
+
+// Configure configures the display with the given options.
+//
+// If no options are given, DisplayDefaults is used.
+func (d *Displays) Configure(opts ...DisplayOption) {
```

Comment:
PTAL -- prototype of the new API.

I updated the PRs between this and the first tests, so you can see [how it's used](https://github.com/tinygo-org/tinygo/pull/497/commits/be8c3ac01e6ab42c14c64f8d61740309da26a01c#diff-f72a4fb4c7cedc6b45bc0df658d1d040R55).

Reply:
This is slightly different from what's proposed above, as I think that the normal closure-based options won't be inlined as well as this version.  Let me know what you think.

Reply:
(trying to reopen on updated push)

Reply:
As you mentioned the closure based API would be very hard to optimize. Basically, there are two reasons why I think a simple config struct would be better:

1. It is used elsewhere in the machine package, and I would like to keep the API consistent.
2. Closures, function pointers, and variadic arguments are all hard to optimize. Not impossible, but hard. Simply using a config struct (like most other APIs in the machine package), even though it may contain lots of fields, is much easier to optimize. Especially when `Configure` is called only once, it is very likely to be optimized away entirely.

My suggestion would be to use a config struct anyway instead of variadic arguments. And if it is too hard/clumsy to use, you could also split the configuration in some way:

```go
interface Display { // just an example
    ConfigureFramebuf(FramebufConfig)
    ConfigureSprites(SpritesConfig)
    ConfigurePalette(PaletteConfig)
}
```

Not sure about the names, but you get the idea.

Reply:
This was basically the thought process behind the original API (that is still visible in PR 7+) -- simple, easily-inlined methods to configure individual pieces of the hardware using namespaced methods.  `machine.Display.Mode3.Configure`, `machine.SpritePalette.Full.Init`, `machine.Tiles.Block[1].Init`, `machine.Background.Offset[1].X.Set` etc.

In this PR, `Display` (originally `Display.Mode3`) has the methods required to be a `Displayer`, but using it in that way makes it pretty non-orthogonal to the rest of what the `machine` package should probably look like.  I still like the original API and would humbly suggest that an /x/drivers API could be made for beginner GBA wrappers, but I defer to you guys as the maintainers to set what kind of look and feel and consistency you want, particularly as I haven't spent any time working with the other machine packages.

Reply:
The thing that is not like other parts of the machine API is the nesting of structs. I would prefer just `machine.<something>.Configure` like all the other APIs. This `<something>` could in fact be `Mode3Display`, `DisplayMode3`, `FramebufDisplay`, or something like that. Or `FullSpritePalette`, etc.